crates/rpc/rpc/src/eth/api/server.rs

```diff
@@ -180,11 +181,84 @@ where
 
     async fn fee_history(
         &self,
-        _block_count: U256,
-        _newest_block: BlockNumber,
+        block_count: u64,
+        newest_block: BlockNumber,
         _reward_percentiles: Option<Vec<f64>>,
     ) -> Result<FeeHistory> {
-        Err(internal_rpc_err("unimplemented"))
+        if block_count == 0 || newest_block < block_count {
+            return Ok(FeeHistory::default())
+        }
+
+        let start_block = newest_block - block_count;
+        let end_block = newest_block;
+
+        let mut fee_history_cache = self.fee_history_cache.lock();
+
+        let mut fee_history_cache_items = BTreeMap::new();
+        let mut first_non_cached_block = None;
+        for block in start_block..=end_block {
+            // Check if block exists in cache, and move it to the head of the list if so
+            if let Some(fee_history_cache_item) = fee_history_cache.get(&block) {
+                fee_history_cache_items.insert(block, fee_history_cache_item.clone());
```

Comment:
I don't understand why we're inserting new data here.

Reply:
added more comments, but the general idea is to populate the `fee_history_cache_items` map from both cache and database, and then use it to compose the response